keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1362 @@
+- Ensure the evacuations are not stomped over and periodically deleted. This means that an
+  Evacuation with a DeletionTimestamp is still a valid evacuation.
+
+If the evacuation is no longer needed, the instigator should remove itself from the finalizers.
+The evacuation will then be deleted by the evacuation controller.
+
+#### Evacuation Instigator Finalizer
+To distinguish between instigator and other finalizers, instigators should use the
+`evacuation-instigator` value. For example, the node maintenance instigator would use the
+`maintenance-a.nodemaintenance.kubernetes.io/evacuation-instigator` finalizer.
+
+Instigator finalizers are automatically removed upon pod removal by evacuation controller. If
+the instigator needs to perform additional final tasks before the Evacuation object deletion, it
+should create a second finalizer without the `evacuation-instigator` value.
+
+### Evacuator
+
+[Evacuee and Evacuator](#evacuee-and-evacuator) section provides a general overview.
+
+There can be multiple evacuators for a single pod, which can be given control of the evacuation by
+the evacuation controller.
+
+We can distinguish 4 different kinds of evacuators:
+- `partial`: it handles only a part of the evacuation process (e.g. releasing one kind of resource).
+  This evacuator should not terminate the pod.
+- `knowledgeable`: has more knowledge (e.g. specific to the deployed application) about how to
+  handle a more graceful evacuation than the controller of the pod (e.g. StatefulSet, Deployment).
+  This can result in the termination of the pod.
+- `controller`: handles full graceful evacuation in a generic way and should result in a pod
+  termination. It can be preempted by a `knowledgeable` evacuator.
+- `fallback`: can handle evacuation if all previous evacuators have failed. This means that
+  `knowledgeable` or `controller` evacuators have either reported no progress, or are simply not
+  implemented for the target pod.
+
+The evacuator should first mark the pods it is interested in evacuating (either partially or fully)
+with the `${EVACUATOR_CLASS}/evacuator-class-priority: ${PRIORITY}/${ROLE}` annotation. This
+annotation is parsed into the `Evacuator` type in the [Evacuation API](#evacuation-api)
+
+- `EVACUATOR_CLASS`: should be unique evacuator's DNS subdomain
+- `PRIORITY` and `ROLE`
+  - `controller` should always set a `PRIORITY=10000` and `ROLE=controller`
+  - other evacuators should set `PRIORITY` according to their own needs. They can use the
+    `controller` evacuator as a reference point, if they want to be run before or after the
+    `controller` evacuator. They can also observe pod annotations and detect what other evacuators
+    have been registered for the evacuation process. `ROLE` is optional and can be used as a signal
+    to other evacuators. The `controller` value is reserved for pod controllers, but otherwise
+    there is no guidance on how the third party evacuators should name their role.
```

Comment:
I have a big concern here.

This design forces a reservation into every public key prefix. I don't think the Kubernetes project should encourage that pattern. It sets a precedent where maybe someone tries this:
```yaml
    sensitive-workload-operator.fruit-company.com/pod-priority-override: "42"
    deployment.apps.kubernetes.io/pod-priority-override: "1000"
    fallback-evacuator.rescue-company.com/pod-priority-override: "2000"
```
only to find that `cloud-provider.example/pod-priority-override: "true"` is already a thing.

Reply:
Yeah, it is not a good pattern. I will try to think of a better `key: value` scheme.

I also thought about adding this to the spec or the status. The problems with that:
- spec: should be static, not dynamic, since the evacuators can come and go during the pod's lifetime
- status: I am not sure if we want to give multiple third party entities access and ask them to update the pod status.

Reply:
Changed to a new format

 ```
evacuation.alpha.coordination.k8s.io/priority-${PRIORITY}: ${EVACUATOR_CLASS}/${ROLE}
```

which would result in

```yaml
  annotations:
    evacuation.alpha.coordination.k8s.io/priority-2000: fallback-evacuator.rescue-company.com
    evacuation.alpha.coordination.k8s.io/priority-10000: deployment.apps.kubernetes.io/controller
    evacuation.alpha.coordination.k8s.io/priority-11000: sensitive-workload-operator.fruit-company.com/knowledgeable-app-specific

```

A disadvantage is that it looks less readable now.

Reply:
Ouch. I am not keen, to put it mildly.

- What happens if two things want to use the same priority number?
- Also, using “alpha” or “beta” in annotation keys is deprecated (so we shouldn't introduce a new one).

Reply:
Thanks, I have removed `alpha` from the annotations and changed them to yet another format.

Unfortunately the key has to be quite long to support all these requirements:

```
evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}: ${PRIORITY}/${ROLE}
```

which would result in

```yaml
  annotations:
    evacuation.coordination.k8s.io/priority_fallback-evacuator.rescue-company.com: "2000"
    evacuation.coordination.k8s.io/priority_deployment.apps.kubernetes.io: "10000/controller"
    evacuation.coordination.k8s.io/priority_sensitive-workload-operator.fruit-company.com: "11000/knowledgeable-app-specific"
```

Reply:
I have started using this format in finalizers as well

```
evacuation.coordination.k8s.io/instigator_${EVACUATION_INSTIGATOR_SUBDOMAIN}
```

which would result in 

```
evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
```

Reply:
I have added a new section  `Length Limitations for Pod Annotations and Evacuation Finalizers`.

I wonder if should implement the hash approach at least for the annotations? The finalizers would be a bit overkill IMO, as there should be only a handful of evacuation instigators.